### PR TITLE
Update ThroughputDistribution with new fields for ThreadPoolControlle…

### PR DESCRIPTION
**Pull request by gjdeval**
…r changes

### CODE REVIEW

File: dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThroughputDistribution.java [hunk 1]
```diff
@@ -47,22 +47,74 @@
...
+
+    /**
+     * The last controller cycle when this distribution was updated. The controller
+     * will discard old data.
+     * This is an int - at 1500ms controller cycles, that allows for 100+ years of
+     * continuous controller operation (no restarts) before it would recycle.
+     */
+    int lastUpdate = 0;
+
+    synchronized void setLastUpdate(int update) {
```
Comment (njr-11) [inline dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThroughputDistribution.java]: method does not appear to be used

File: dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThroughputDistribution.java [hunk 2]
```diff
@@ -47,22 +47,74 @@
...
+        lastUpdate = update;
+    }
+
+    synchronized int getLastUpdate() {
```
Comment (njr-11) [inline dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThroughputDistribution.java]: method does not appear to be used

**Comment by njr-11** [comment]
Gary - I added Tim to the list of reviewers.  His team typically works with the internals of the Liberty thread pool.  I can look over the code, but haven't worked in this part of the thread pool before, so I want to make sure you are getting the right level of expertise for this.

File: dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThroughputDistribution.java (hunk 2 above)
Comment (gjdeval) [reply dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThroughputDistribution.java]: Thanks for the feedback Nathan. 
Methods in this class are used by ThreadPoolController, which is in the same package. getLastUpdate() is used by the changes I'm making now in ThreadPoolController, which should be committed in the next few days.

**Comment by kaczyns** [comment]
Should I be reviewing this now, or waiting for something else here?

**Comment by gjdeval** [comment]
My original plan was to deliver ThroughputDistribution first, since it did not depend on my changes in ThreadPoolController (but ThreadPoolController does depend on the changes in ThroughputDistribution), and I thought the review/approval would be very quick, and I was not ready to drop the ThreadPoolController changes ...
But now ThreadPoolController (and ExecutorServiceImpl) are very nearly ready, just making some final checks, so I think that I'll just add them to this PR, so it all goes thru in one bundle ... 
So you can hold off on this review, the whole package should arrive with a thud shortly.

**Comment by gjdeval** [comment]
OK, now I've added ExecutorServiceImpl (small changes to allow an often-used field in ThreadPoolController to be final) and ThreadPoolController (*extensive* changes to implement the redesign described in the POC package). Thanks very much for reviewing it all.

**Comment by vijaysun-omr** [comment]
https://github.com/gjdeval/open-liberty/blob/2b9492a6f3a3b7e3d10863e954550f8c05d09a8b/dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java#L180

...the code (and comment) suggest that we will retry the request (for setting a particular thread pool size). Is it worth considering abandoning the request after some number of attempts ?

**Comment by vijaysun-omr** [comment]
https://github.com/gjdeval/open-liberty/blob/2b9492a6f3a3b7e3d10863e954550f8c05d09a8b/dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java#L195

...could the comment for this be beefed up to clarify why a combination of age and standard deviation is relevant for the pruning decision.

**Comment by vijaysun-omr** [comment]
https://github.com/gjdeval/open-liberty/blob/2b9492a6f3a3b7e3d10863e954550f8c05d09a8b/dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java#L1488

...so the pruning is only done for thread pool sizes outside of a certain compare range multiple from the current thread pool, is that correct ? i.e. not all "aged" thread pool sizes' data is pruned.

**Comment by vijaysun-omr** [comment]
https://github.com/gjdeval/open-liberty/blob/2b9492a6f3a3b7e3d10863e954550f8c05d09a8b/dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java#L222

..."ratio-based grow/shrink considerations" is really the leanTowardShrinking logic and related to poolTputRatioHigh and poolTputRatioLow variables that were declared just prior to this line. Maybe make that more explicit in the comment.

**Comment by vijaysun-omr** [comment]
https://github.com/gjdeval/open-liberty/blob/2b9492a6f3a3b7e3d10863e954550f8c05d09a8b/dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java#L731

How can the variable throughput (i.e. the observed throughput) be < 0.0 ?

**Comment by vijaysun-omr** [comment]
https://github.com/gjdeval/open-liberty/blob/2b9492a6f3a3b7e3d10863e954550f8c05d09a8b/dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java#L733

calls resetStatistics whereas 

https://github.com/gjdeval/open-liberty/blob/2b9492a6f3a3b7e3d10863e954550f8c05d09a8b/dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java#L795

calls resetThreadPool. I know that resetThreadPool calls resetStatistics but I wanted to ask why the choice of the reset method is different at the above 2 spots in the code.

**Comment by gjdeval** [comment]
Not sure at all, that check pre-dates my involvement ... assume it is some
peculiar corner-case protection.


Cheers,
Gary DeVal
IBM WebSphere Performance Engineering
Email:  gjdeval@us.ibm.com

**Comment by vijaysun-omr** [comment]
@gjdeval Thanks for the changes and answers. I am fine with this getting merged subject to @jhanders34 review being done.

**Comment by gjdeval** [comment]
#build

**Comment by jhanders34** [comment]
#build

**Comment by gjdeval** [comment]
#build

**Comment by gjdeval** [comment]
#build

**Comment by gjdeval** [comment]
#build

**Comment by jhanders34** [comment]
#build

**Pull request merged**